KAFKA-17710; Rework uniform heterogeneous assignor to improve perf

squah-confluent opened the pull request:
KAFKA-17710; Rework uniform heterogeneous assignor to improve perf

Rework the uniform heterogeneous assignor to improve performance, while
preserving the high level ideas and structure from the existing
implementation:
 * The assignor works in 3 stages: importing the previous assignment for
   stickiness, assigning unassigned partitions and iteratively
   reassigning partitions to improve balance.
 * Unassigned partitions are assigned to the subscribers with the least
   number of partitions. This maximizes balance within a single topic.
 * During the iterative rebalancing phase, partitions are reassigned to
   their previous owner if it improves balance (stickiness restoration).
 * During the iterative rebalancing phase, partitions are reassigned to
   the subscriber with the least number of partitions to improve
   balance.

A non-exhaustive list of changes is:
 * The assignment of unassigned partitions and iterative reassignment
   stages now works through partitions topic by topic. Previously
   partitions from topics with the same number of partitions per
   subscriber would be interleaved. Since we iterate topic by topic, we
   can reuse data about topic subscribers.
 * Instead of maintaining TreeSets to find the least loaded subscribers,
   we sort an ArrayList of subscribers once per topic and start filling
   up subscribers, least loaded first. In testing, this approach was
   found to be faster than maintaining PriorityQueues.
 * Implement stickiness restoration by creating a mapping of partitions
   to previous owner and checking against that mapping, instead of
   tracking partition movements during iterative reassignment.
 * Track member partition counts using a plain int array, to avoid
   overhead from boxing and HashMap lookups. Member partition counts are
   accessed very frequently and this needs to be fast. As a consequence,
   we have to number members 0 to M - 1.
 * Bound the iterative reassignment stage to a fixed number of
   iterations. Under some uncommon subscription patterns, the iterative
   reassignment stage converges slowly. In these cases, the iterative
   reassignment stage terminates without producing an optimally balanced
   assignment anyway (see javadoc for balanceTopics).
 * Re-use Maps from the previous assignment where possible,
   ie. introduce a copy-on-write mechanism while computing the new
   assignment.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

squah-confluent committed:
KAFKA-17710; Rework uniform heterogeneous assignor to improve perf

Rework the uniform heterogeneous assignor to improve performance, while
preserving the high level ideas and structure from the existing
implementation:
 * The assignor works in 3 stages: importing the previous assignment for
   stickiness, assigning unassigned partitions and iteratively
   reassigning partitions to improve balance.
 * Unassigned partitions are assigned to the subscribers with the least
   number of partitions. This maximizes balance within a single topic.
 * During the iterative rebalancing phase, partitions are reassigned to
   their previous owner if it improves balance (stickiness restoration).
 * During the iterative rebalancing phase, partitions are reassigned to
   the subscriber with the least number of partitions to improve
   balance.

A non-exhaustive list of changes is:
 * The assignment of unassigned partitions and iterative reassignment
   stages now works through partitions topic by topic. Previously
   partitions from topics with the same number of partitions per
   subscriber would be interleaved. Since we iterate topic by topic, we
   can reuse data about topic subscribers.
 * Instead of maintaining TreeSets to find the least loaded subscribers,
   we sort an ArrayList of subscribers once per topic and start filling
   up subscribers, least loaded first. In testing, this approach was
   found to be faster than maintaining PriorityQueues.
 * Implement stickiness restoration by creating a mapping of partitions
   to previous owner and checking against that mapping, instead of
   tracking partition movements during iterative reassignment.
 * Track member partition counts using a plain int array, to avoid
   overhead from boxing and HashMap lookups. Member partition counts are
   accessed very frequently and this needs to be fast. As a consequence,
   we have to number members 0 to M - 1.
 * Bound the iterative reassignment stage to a fixed number of
   iterations. Under some uncommon subscription patterns, the iterative
   reassignment stage converges slowly. In these cases, the iterative
   reassignment stage terminates without producing an optimally balanced
   assignment anyway (see javadoc for balanceTopics).
 * Re-use Maps from the previous assignment where possible,
   ie. introduce a copy-on-write mechanism while computing the new
   assignment.

squah-confluent committed:
fixup: fix edge case

squah-confluent commented:
Before benchmarks taken at commit c8cfb4c7f1ff906cc3ac0d22ba9b99c0c8a14e9c

```
assignorType: UNIFORM
isRackAware: false
subscriptionType: HETEROGENEOUS

                                                                                                                                         Before    After
Benchmark                                       (assignmentType)  (memberCount)  (partitionsToMemberRatio)  (topicCount)  Mode  Cnt       Score    Score  Units
ServerSideAssignorBenchmark.doAssignment                    FULL            100                          5            10  avgt    2       1.432    0.059  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            100                          5           100  avgt    2       1.644    0.233  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            100                          5          1000  avgt    2       1.789    1.648  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            100                         10            10  avgt    2       3.165    0.076  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            100                         10           100  avgt    2       3.208    0.271  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            100                         10          1000  avgt    2       4.751    2.378  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            100                         50            10  avgt    2      15.248    0.261  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            100                         50           100  avgt    2      15.477    0.445  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            100                         50          1000  avgt    2      18.254    2.751  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                          5            10  avgt    2      41.499    0.283  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                          5           100  avgt    2      48.283    0.775  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                          5          1000  avgt    2      44.881    8.025  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                         10            10  avgt    2      98.130    0.428  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                         10           100  avgt    2      91.082    1.159  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                         10          1000  avgt    2     104.599    8.276  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                         50            10  avgt    2     452.457    1.200  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                         50           100  avgt    2     451.165    2.248  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL            500                         50          1000  avgt    2     446.059   10.350  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                          5            10  avgt    2     241.544    0.577  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                          5           100  avgt    2     212.019    1.574  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                          5          1000  avgt    2     233.933   15.169  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                         10            10  avgt    2     411.058    0.906  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                         10           100  avgt    2     396.443    2.077  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                         10          1000  avgt    2     440.244   15.598  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                         50            10  avgt    2    2056.629    3.113  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                         50           100  avgt    2    2131.425    4.621  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           1000                         50          1000  avgt    2    2289.710   20.684  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                          5            10  avgt    2    6720.876    3.023  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                          5           100  avgt    2    7314.790    8.132  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                          5          1000  avgt    2    7332.188   96.327  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                         10            10  avgt    2   12004.132    4.261  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                         10           100  avgt    2   12875.358   10.446  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                         10          1000  avgt    2   14477.062   93.610  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                         50            10  avgt    2   64960.501   15.259  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                         50           100  avgt    2   58660.843   24.307  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL           5000                         50          1000  avgt    2   72814.844  122.486  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                          5            10  avgt    2   31590.175    5.790  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                          5           100  avgt    2   29001.639   18.114  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                          5          1000  avgt    2   31803.961  202.708  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                         10            10  avgt    2   65327.922    8.923  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                         10           100  avgt    2   68236.626   22.947  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                         10          1000  avgt    2   64831.264  190.180  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                         50            10  avgt    2  278313.470   31.564  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                         50           100  avgt    2  279095.078   52.117  ms/op
ServerSideAssignorBenchmark.doAssignment                    FULL          10000                         50          1000  avgt    2  314830.161  241.265  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                          5            10  avgt    2       0.424    0.050  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                          5           100  avgt    2       0.544    0.169  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                          5          1000  avgt    2       1.850    1.674  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                         10            10  avgt    2       0.816    0.064  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                         10           100  avgt    2       0.915    0.194  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                         10          1000  avgt    2       2.596    1.845  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                         50            10  avgt    2       4.053    0.178  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                         50           100  avgt    2       4.174    0.358  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            100                         50          1000  avgt    2       6.115    2.149  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                          5            10  avgt    2       2.661    0.235  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                          5           100  avgt    2       3.332    0.685  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                          5          1000  avgt    2      11.108    6.851  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                         10            10  avgt    2       4.916    0.292  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                         10           100  avgt    2       5.906    0.815  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                         10          1000  avgt    2      14.924    6.980  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                         50            10  avgt    2      23.836    0.788  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                         50           100  avgt    2      24.090    1.584  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL            500                         50          1000  avgt    2      36.857    9.050  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                          5            10  avgt    2       5.641    0.460  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                          5           100  avgt    2       7.331    1.550  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                          5          1000  avgt    2      22.841   15.560  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                         10            10  avgt    2       9.969    0.537  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                         10           100  avgt    2      12.382    1.619  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                         10          1000  avgt    2      28.463   14.274  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                         50            10  avgt    2      45.711    1.497  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                         50           100  avgt    2      50.599    3.115  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           1000                         50          1000  avgt    2      71.140   18.617  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                          5            10  avgt    2      31.629    2.372  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                          5           100  avgt    2      40.474    7.559  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                          5          1000  avgt    2     142.004   82.413  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                         10            10  avgt    2      62.571    2.793  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                         10           100  avgt    2      71.818    8.975  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                         10          1000  avgt    2     163.790   81.646  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                         50            10  avgt    2     294.250    7.747  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                         50           100  avgt    2     311.555   16.663  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           5000                         50          1000  avgt    2     407.648   91.072  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                          5            10  avgt    2      63.838    5.053  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                          5           100  avgt    2      88.044   16.175  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                          5          1000  avgt    2     262.618  159.226  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                         10            10  avgt    2     124.435    6.151  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                         10           100  avgt    2     151.501   19.379  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                         10          1000  avgt    2     344.053  162.744  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                         50            10  avgt    2     626.251   14.456  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                         50           100  avgt    2     761.011   33.645  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL          10000                         50          1000  avgt    2     909.380  187.965  ms/op
```

squah-confluent committed:
fixup: fix typo

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -75,79 +84,153 @@ public class UniformHeterogeneousAssignmentBuilder {
+
+        // Number the members 0 to M - 1.
+        this.memberIds = new ArrayList<>(groupSpec.memberIds());
+        this.memberIds.sort(null);
```
Is there a reason for sorting the member ids here?

squah-confluent replied:
It's entirely for determinism.

dajac replied:
Is it required by the algorithm or is it only required to make the tests deterministic?

squah-confluent replied:
It's not required by the algorithm. It's purely for the unit tests to have deterministic output.

dajac replied:
Would it be possible to remove it and to ensure that the member are sorted in the unit test?

squah-confluent replied:
Yes, I've removed it and added a GroupSpec implementation in the unit tests that returns memberIds in sorted order.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -75,79 +84,153 @@ public class UniformHeterogeneousAssignmentBuilder {
+        };
+
+        // Initialize partition owners for the previous and target assignments.
+        this.previousAssignmentPartitionOwners = new HashMap<>((int) ((this.subscribedTopicIds.size() / 0.75f) + 1));
```
In the GroupSpec, we have the isPartitionAssigned(topicId, partitionId) method which can be used to determine if a partition is already assigned in the previous target assignment. Have you considered re-using it? Or, is there a reason not to reuse it?

squah-confluent replied:
`previousAssignmentPartitionOwners` is only used during the iterative rebalance phase to restore partitions to their previous owners if it helps balance. To use `GroupSpec`, we'd need to update the method or add a new method to return the assigned memberId. Then compared to the current approach, we'd end up with an extra Uuid map lookup and an extra memberId map lookup per query. I think there would be a performance cost to this, but I have not tested it.

I'm actually tempted to remove the usage of this entirely. See comments on `balanceTopicRestoringStickiness`.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -75,79 +84,153 @@ public class UniformHeterogeneousAssignmentBuilder {
+     * </ol>
      */
     public GroupAssignment build() {
         if (subscribedTopicIds.isEmpty()) {
```
nit: Not related but we could remove the info below.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+
+                    int numPartitions = subscribedTopicDescriber.numPartitions(topicId);
+                    for (int partition : partitions) {
+                        if (partition >= numPartitions) {
```
AFAIK Kafka does not support reducing the number of partitions of a topic hence I wonder if this check is necessary. For instance, we don't have it in the heterogenous case.

squah-confluent replied:
I think we should remove the check for consistency with the homogeneous case.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+                            }
+                            memberTargetAssignmentSizes[memberIndex]--;
+
+                            continue;
```
nit: I am not a fan of this pattern. Would it be possible to use a if/else instead? Unless if it performs better this way, I would prefer it. If it performs better, let's add a small comment about it.

squah-confluent replied:
Now that the check for reduced partition counts is gone, this code no longer exists.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+     * Imports the existing assignment into the target assignment, taking into account topic
+     * unsubscriptions.
      */
     private void assignStickyPartitions() {
```
nit: It seems that this method does not really assign partitions anymore but rather revoke the unnecessary partitions. Should we name it something like `maybeRevokePartitions`?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
         assignStickyPartitions();
 
-        unassignedPartitionsAssignment();
+        computeUnassignedPartitions(unassignedPartitions);
```
What's the reason for passing `unassignedPartitions` as an argument to the method vs accessing it directly from the method? It seems that the method is used only once.

squah-confluent replied:
It's a half-done refactor from the previous implementation where unassignedPartitions was used in more than one place. I think it should be turned into a local variable here.

squah-confluent replied:
The current code looks like
```java
        Map<Uuid, List<Integer>> unassignedPartitions = computeUnassignedPartitions();
        assignRemainingPartitions(unassignedPartitions);
```
What are your thoughts on it?

dajac replied:
That works for me. Thanks.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+     * @param unassignedPartitions          The map in which to store the unassigned partitions.
+     */
+    private void computeUnassignedPartitions(Map<Uuid, List<Integer>> unassignedPartitions) {
+        unassignedPartitions.clear();
```
Do we really need to call `clear()`?

squah-confluent replied:
It's not strictly necessary. After turning `unassignedPartitions` into a local var it's no longer present.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
 
-        unassignedPartitionsAssignment();
+        computeUnassignedPartitions(unassignedPartitions);
+        unassignedPartitionsAssignment(unassignedPartitions);
```
Same question about `unassignedPartitions` here.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
      */
-    private boolean canMemberParticipateInReassignment(String memberId) {
-        Set<Uuid> assignedTopicIds = targetAssignment.get(memberId).partitions().keySet();
+    private void unassignedPartitionsAssignment(Map<Uuid, List<Integer>> partitions) {
```
nit: assignRemainingPartitions or assignUnassignedPartitions?

squah-confluent replied:
Renamed it to assignRemainingPartitions for consistency with the homogeneous assignor.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
-            .thenComparingInt(topicIdPartition -> membersPerTopic.get(topicIdPartition.topicId()).size())
-            .thenComparingInt(TopicIdPartition::partitionId);
+    private List<Uuid> sortTopicIds(Collection<Uuid> topicIds) {
+        Comparator<Uuid> comparator = new Comparator<Uuid>() {
```
Is it worth declaring this comparator once as a class attribute? It seems that `sortTopicIds` is called multiple times so we don't have to re-instanciate it.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+         * @return The difference in partition counts between the most and least loaded members.
+         */
+        public int initialize(List<Integer> members) {
+            if (members.size() < 1) {
```
nit: isEmpty()?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
-            return true;
+            for (int partition : partitions.get(topicId)) {
+                int leastLoadedMemberIndex = memberAssignmentBalancer.nextLeastLoadedMember();
 
```
nit: We could remove this empty line. It does not bring much.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+
+        // Repeat reassignment until no partition can be moved to improve the balance or we hit an
+        // iteration limit.
+        for (int i = 0; i < 16; i++) {
```
Should we define a constant for 16? It may also be good to explain why we picked 16.

squah-confluent replied:
Yes. The choice of 16 was arbitrary. I didn't want something too low (not enough time for convergence) or too high (too much cpu cost).

Also added a javadoc for the new constant.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+        for (int i = 0; i < numPartitions; i++) {
+            int partition = partitions.get(i);
+            int ownerIndex = partitionOwners[partition];
+            if (startPartitionIndices.get(ownerIndex) == null) {
```
nit: putIfAbsent?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+     * @param partitionOwners          The current owners of the partitions.
+     * @return the number of partitions reassigned.
+     */
+    private int balanceTopicRestoringStickiness(
```
I wonder if this step really make a big different. Does it based on your experience?

squah-confluent replied:
This step is a rough translation of something the previous implementation did. The only time it does anything is if a previous balancing iteration moved partitions away from a member and now wants to move them back. For this to happen, subscriptions must be overlapping and non-equal.

eg. Consider a case with 2 topics and 3 consumers, where topic 1 is subscribed to by consumer 1 and 2 and topic 2 is subscribed to by consumer 2 and 3. Let's say that consumer 1 has 10 partitions, 2 has 15 and 3 has 5. We might balance topic 1 first and move partitions from consumer 2 -> 1, so that they both have 12-13 partitions. Then we might balance topic 2 and move partitions from consumer 2 -> 3, so that they both have 9 partitions. The next iteration on topic 1 would move partitions back from consumer 1 -> 2.

Since subscription patterns like these are uncommon, I'm actually tempted to remove this step. What do you think?

dajac replied:
Yeah, I agree. This was my first though too.

dajac replied:
Do you have an idea of the gain if we remove it?

squah-confluent replied:
I took a look at profiles for the smallest and largest full and incremental cases and it's pretty small, consistently less than 0.50%.

dajac replied:
Ack. I am fine with removing it if it is not really useful in the end.

squah-confluent replied:
I've removed it and `previousAssignmentPartitionOwners` in the latest version.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
+        memberTargetAssignmentSizes[memberIndex]--;
+    }
+
+    private static Map<Uuid, Set<Integer>> deepCopy(Map<Uuid, Set<Integer>> map) {
```
We have the very same method in UniformHomogeneousAssignmentBuilder. I wonder if we could share them. We could also potentially share isImmutableMap too. One way would be to have a small class of helpers. Have you considered it?

squah-confluent replied:
I've factored these out into an `AssignorHelpers` class.

dajac reviewed (commented):
@squah-confluent Thanks for the patch. I just made a first pass on it and I left some comments for consideration.

squah-confluent committed:
fixup: remove info log for empty subscription list

squah-confluent committed:
fixup: removal partition count reduction check

squah-confluent committed:
fixup: convert unassignedPartitions to a local variable

squah-confluent committed:
fixup: rename unassignedPartitionsAssignment to assignRemainingPartitions

squah-confluent committed:
fixup: rename assignStickyPartitions to maybeRevokePartitions and sync with UniformHomogeneneousAssignmentBuilder

squah-confluent committed:
fixup: hold on to topic comparator

squah-confluent committed:
fixup: s/.size()/.isEmpty()/

squah-confluent committed:
fixup: remove newline

squah-confluent committed:
fixup: use putIfAbsent instead of get then put

squah-confluent committed:
fixup: add MAX_ITERATION_COUNT constant

squah-confluent committed:
fixup: factor out AssignorHelpers class

squah-confluent committed:
fixup: actually add the AssignorHelpers class

squah-confluent committed:
fixup: make spotless happy

squah-confluent committed:
Merge commit 'b03fe66cfe3b0457c856b12772ac4a95c6eedbbb' into squah-rework-uniform-heterogeneous-assignor

squah-confluent committed:
fixup: s/mkSet/Set.of/

eschwabe committed:
Merge pull request #173 from confluentinc/squah-rework-uniform-heterogeneous-assignor

merge "KAFKA-17759 Remove Utils.mkSet" and update test code

squah-confluent committed:
fixup: remove balanceTopicRestoringStickiness

Remove balanceTopicRestoringStickiness since it only benefits uncommon
subscription patterns. Even under those uncommon subscription patterns,
the common case of a single member leaving or joining does not cause
the step to do anything.

squah-confluent committed:
fixup: remove sort of member ids

squah-confluent committed:
fixup: add GroupSpec implementation with deterministic memberIds in tests

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java:
```diff
@@ -75,798 +86,841 @@ public class UniformHeterogeneousAssignmentBuilder {
 
     /**
-     * List of subscribed members for each topic.
+     * The list of members in the consumer group, sorted by member id.
```
nit: Should we remove the "sorted by member id"?

squah-confluent replied:
Good spot, I've updated the javadoc.

dajac reviewed (commented):
@squah-confluent Thanks for the update. I just made another pass on it and I left one small nit. I think that we can merge it when addressed.

squah-confluent committed:
fixup: fix javadoc

dajac reviewed (approved):
LGTM, thanks for the patch!
